clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/AbstractAtLeastOnceKafkaConsumer.java

```diff
@@ -256,43 +257,83 @@ private void handleBatch(final KafkaConsumerRecords<String, Buffer> records) {
         try {
             final T message = createMessage(record);
             messageHandler.handle(message);
+            addToCurrentOffsets(record);
             return Future.succeededFuture();
         } catch (final Exception ex) {
             LOG.error("Error handling record, closing the consumer: ", ex);
-            commitCurrentOffset(record); // prevents that already processed records from the batch are consumed again
-            stop();
+            tryCommitAndClose();
             return Future.failedFuture(ex);
```

Comment:
In case this "catch" block getting run, will the failed Future here cause the block with "LOG.error("Error polling messages: " + cause);" inside `poll`  above to be run?
If so, `close` would get called twice.
If no, the `closeHandler` wouldn't get called here.

Reply:
I don't see a problem here. Error handling of the `poll` operation should not be reached. And it is intended that that close handler is not invoked if the message handler failed (the code that uses this class sets both handlers, so it "knows" that it threw an exception).

Reply:
> so it "knows" that it threw an exception

If it's an unexpected exception and there's no try/catch inside the message handler, the message handler might not know. I think it would be more consistent, to also call the close handler then. And the "onClose" logic would only need to be in one place.

In any case, the class javadoc currently says, that the close handler is called:
````
 * If the message processing fails because either {@link #createMessage(KafkaConsumerRecord)} or the message handler
 * throws an exception, the Kafka consumer will be closed and the exception will be passed to the close handler.
````

Reply:
> If it's an unexpected exception and there's no try/catch inside the message handler, the message handler might not know. I think it would be more consistent, to also call the close handler then. And the "onClose" logic would only need to be in one place.

These are valid points. On the other hand: how would the caller handle the unexpected exception? The generic failure handling could be to start a new consumer. That would lead to the same exception, causing a failure loop...

Initially, it was like that. @sophokles73 seem to have expected the opposite: https://github.com/eclipse/hono/pull/2432#discussion_r565301540
Therefore I changed it (and obviously missed to adapt this point in that place of the JavaDoc). 
FMPOV both variants are valid. Of course, the documentation must match the behavior.

Reply:
> On the other hand: how would the caller handle the unexpected exception? The generic failure handling could be to start a new consumer. 

You mean, if there's a try/catch inside the message handler, what to do in the "catch" block? I don't see a reason for starting a new consumer there. When an exception happens there, the issue is with the record or some other consumer independent state, not the consumer itself, isn't it? And besides, actually starting a new consumer there while the current consumer is still alive would be error-prone.
So, this means that I think it's not very practical to close the consumer when there is an exception in `createMessage` or `messageHandler.handle(message);`. I would rather invoke a `onMessageHandlingError(record)` method which by default just logs the error. That method can then be overridden by actual Consumer implementations if needed.

Reply:
Since using the `AbstractAtLeastOnceKafkaConsumer` means having to implement the closeHandler anyway (to handle commit errors, etc.), I would also be ok with closing the consumer on messageHandler errors (which should be seen as programming errors). But then I think indeed the closeHandler should be called there.
And I think it would be good to document that the messageHandler shouldn't explicitly throw an exception (because otherwise the consumer gets closed).

Reply:
@calohmn I agree, it would be better to invoke the close handler on every kind of error. Otherwise, at least for programming errors, the consumer would be closed silently. And it would improve the testability of the class if I could see from the outside when the consumer closes.

Reply:
@sophokles73 Invoking the close handler on errors during message consumption would also affect the behavior of the Kafka-based downstream consumers returned by the application client factory. I would change the documentation in https://github.com/eclipse/hono/blob/82a600c770b68e719c429ec1c4cc8f8f49a4a09e/clients/application-kafka/src/main/java/org/eclipse/hono/application/client/kafka/KafkaApplicationClientFactory.java#L46-L47 accordingly. Are you OK with that?

Reply:
>  Invoking the close handler on errors during message consumption would also affect the behavior of the Kafka-based downstream consumers returned by the application client factory.

Are we talking about exceptions being thrown accidentally or exceptions thrown deliberately by the consumer to indicate that it cannot/will not process the message? I guess the latter behavior should be ruled out explicitly (i.e. be documented properly).

Reply:
Yes, I agree.

Reply:
So then this is indeed something that we want to prevent the message handler from doing. That is in contrast to the AMQP based consumer where we explicitly allow the user code to throw an (unchecked) exception to indicate that the message should be settled with the rejected/released outcome. Allowing it there but preventing it here will probably make it much harder to reach the goal of implementing a consumer once and use it for both AMQP and Kafka ...

Reply:
Yes, I don't really see a way around that. AMQP allows rejecting a single message. Kafka's streaming semantics do not provide this. You can ignore a message (which could break the AT LEAST ONCE guarantee) or you can stop the stream and wait for help :-) So, if one wants to use Kafka-based messaging one should accept all messages and do not rely on a "backchannel" to the sender. 

I would say that the message handler *should* not throw an exception. Because this would stop the consumption. If users of the consumer decide to throw an exception in the message handler nevertheless, they get their consumer closed and need to handle them in the close handler. Do you have a better idea?

Reply:
Maybe we can use the MessageContext to provide an abstraction for the consumer in this case as well. I can imagine to add *MessageContext.accept()*, *reject()* and *release()* methods which for AMQP have the expected consequences regarding the disposition update and for Kafka may or may not have an impact on which index is being committed or the Kafka consumer being closed.

Reply:
The question remains what the Kafka consumer should do to reject or release a message. 

FMPOV the goal is not that consumers for AMQP and Kafka behave exactly the same but to have an API that makes it easy to switch from AMQP to Kafka. It should also allow it to implement a solution that supports both variants (which might involve type checking and casting when things should be handled differently).
The alternative is to aim for the lowest common denominator. In this case, this would mean to refrain from transmitting the status to the sender.

Reply:
I pushed the change that invokes the close handler if the message hanlder throws an exception. And I changed the JavaDoc in the northbound application client accordingly. @sophokles73 Does this make sense from your POV?

Reply:
I am still missing support for a very simple scenario that I believe to be quite common:
When the user provided consumer receives a message for processing it might want to retrieve some data from a database or a remote service first. Let's say that step fails with a runtime exception. My understanding is that the consumer needs to implement retry behavior itself in this case, right? In the AMQP case the runtime exception would trigger re-delivery of the (event) message and thus (implicitly) re-run the consumer with the message (AT LEAST ONCE). In the Kafka case, however, the client will close the Kafka Consumer and notify the close handler with the runtime exception. Alternatively, the consumer can catch the RE itself and re-run its handling logic. When the client crashes before it has committed an already processed message, however, it will get the message again for processing after a restart. Is this how it is supposed to work or am I missing something?

Reply:
> Alternatively, the consumer can catch the RE itself and re-run its handling logic. 

That would probably be the preferred way. But if this takes very long, it could also close the consumer (either by invoking the `stop()` method or by throwing an exception) and have the message re-delivered when it starts a new consumer.

> Is this how it is supposed to work 

Yes. Is there anything that should be added?

Reply:
> I am still missing support for a very simple scenario that I believe to be quite common:

I'm afraid I don't understand what you're missing. Do you think we should implement an automatic restart in the consumer?

Reply:
I was referring to the automatic replay of messages as in the AMQP case. It still strikes me as odd that the programming model for implementing the consumer is so different between AMQP and Kafka. We could let the Kafka based DownstreamMessageConsumer re-deliver the message automatically to the client provided consumer, if it throws a runtime exception. Then it would be the same behavior as in the AMQP case, wouldn't it?

Reply:
I'll think about it. I guess it wouldn't be necessary to poll the record again from Kafka but instead invoke the message handler with the already created `DownstreamMessage` object. The result will be that if the error is not transient it will infinitely call the message handler and create and throw an exception (and no other event respectively telemetry message will be processed for this tenant). This will happen for NPEs in the message handler, other programming errors, or malformed messages. 
Wouldn't it be better, if we want to go this route, that based on the exception thrown, we can distinguish between expected, potentially transient, errors and unexpected permanent ones?

Reply:
Sounds like a good idea. In any case, I would like to align the behavior of the Kafka and AMQP based clients wrt this as much as (reasonably) possible in order to minimize differences in the programming model.

Reply:
We could then define an exception type that the handler is expected to throw and which indicates that it wants a "re-try". All other exceptions are then unexpected and lead to closing the consumer. WDYT?

Reply:
In the ServiceErrorException hierarchy we already distinguish between client induced and server side induced errors. In the AMQP based client a `ClientErrorException` leads to a `rejected` outcome whereas a `ServerErrorException` leads to a `released` outcome. IMHO we could therefore do a re-tried delivery for `ServerErrorException`s thrown by the consumer as they usually indicate a transient problem (at least in Hono). All other exceptions could then cause the consumer to be closed and the close handler being invoked. Would that work?

Reply:
@sophokles73 Please have a look at my last commit. I added this there.

Reply:
@sophokles73 There is still a difference between `ProtonBasedApplicationClientFactory` and `KafkaApplicationClientFactory` regarding error handling. Do you want to adjust the behavior of the AMQP based client factory as well?

Reply:
IMHO we should never let the client close the Kafka consumer in case of runtime exceptions thrown when processing messages. Can you please change the code and update the JavaDocs accordingly @b-abel ?

Reply:
Sure. We are talking about the message _processing_ step (i.e. the message handler invocation), right? And not about errors in the communication with the Kafka cluster (i.e. _poll_ and _commit_)?

Reply:
yes

Reply:
Done. Please see my last commit.